src/main/java/org/reaktivity/nukleus/http_cache/internal/proxy/cache/DefaultCache.java

```diff
@@ -144,24 +169,96 @@ public boolean matchCacheableRequest(
         short authScope,
         int requestHash)
     {
-        final DefaultCacheEntry cacheEntry = cachedEntries.get(requestHash);
+        final DefaultCacheEntry cacheEntry = cachedEntriesByRequestHash.get(requestHash);
         return satisfiedByCache(requestHeaders) &&
                 cacheEntry != null &&
-                cacheEntry.canServeRequest(requestHeaders, authScope);
-
+                cacheEntry.canServeRequest(requestHeaders, authScope) &&
+                cacheEntry.isValid();
     }
 
     public void purge(
         int requestHash)
     {
-        DefaultCacheEntry cacheEntry = this.cachedEntries.remove(requestHash);
-        entryCount.accept(-1);
+        DefaultCacheEntry cacheEntry = cachedEntriesByRequestHash.remove(requestHash);
         if (cacheEntry != null)
         {
+            final int requestHashWithoutQuery = cacheEntry.requestHashWithoutQuery();
+            Int2ObjectHashMap cachedEntriesByRequestHashFromWithoutQueryList = cachedEntriesByRequestHashWithoutQuery
+                .computeIfPresent(requestHashWithoutQuery, (h, m) -> m.remove(requestHash) != null && m.isEmpty() ? null : m);
+            if (cachedEntriesByRequestHashFromWithoutQueryList == null)
+            {
+                cachedEntriesByRequestHashWithoutQuery.remove(requestHashWithoutQuery);
+            }
+
+            entryCount.accept(-1);
             cacheEntry.purge();
         }
     }
 
+    public void invalidateCacheEntryIfNecessary(
+        HttpCacheProxyFactory factory,
+        int requestHash,
+        String requestURL,
+        ArrayFW<HttpHeaderFW> headers)
+    {
+        DefaultCacheEntry cacheEntry = cachedEntriesByRequestHash.remove(requestHash);
+        if (cacheEntry != null)
+        {
+            cacheEntry.invalidate();
+        }
+
+        headers.forEach(header ->
+        {
+            invalidateLinkCacheEntry(factory, requestURL, header);
+        });
+    }
+
+    private void invalidateLinkCacheEntry(
+        HttpCacheProxyFactory factory,
+        String requestURL,
+        HttpHeaderFW header)
+    {
+        final String linkName = header.name().asString();
+        final String linkValue = header.value().asString();
+        if (LINK.equals(linkName))
+        {
+            for (String linkTarget: linkValue.split("\\s*,\\s*"))
+            {
+                Matcher matcher = LINK_URL_PATTERN.matcher(linkTarget);
+                if (matcher.matches())
+                {
+                    final URI requestURI = URI.create(requestURL);
+                    final String linkTargetScheme = matcher.group("scheme");
+                    final String linkTargetHostname = matcher.group("hostname");
+
+                    if ((linkTargetScheme != null && linkTargetHostname != null) &&
+                        (!requestURI.getScheme().equals(linkTargetScheme) || !requestURI.getHost().equals(linkTargetHostname)))
```

Comment:
Change last part to use `!linkTargetScheme.equals(requestURI.getScheme())`, and `!linkTargetHostname.equals(requestURI.getHost())` to protect against `null` return values from `requestURI` methods, given that you have already checked that `linkTargetScheme` and `linkTargetHostname` are non-null.

Reply:
yep thanks